zcash_client_sqlite/src/wallet.rs

```diff
@@ -129,8 +129,14 @@ pub(crate) fn pool_code(pool_type: PoolType) -> i64 {
 }
 
 pub(crate) fn memo_repr(memo: Option<&MemoBytes>) -> Option<&[u8]> {
-    memo.filter(|m| *m != &MemoBytes::empty())
-        .map(|m| m.as_slice())
+    memo.map(|m| {
+        if m == &MemoBytes::empty() {
+            // we store the empty memo as a single 0xf6 byte
+            &[0xf6]
+        } else {
+            m.as_slice()
+        }
+    })
```

Comment:
The memo may be `None` in the `sent_notes` table as a consequence of wallet recovery; the note may have been detected but the rest of the transaction data will not have been populated. The intent of this change overall is to make `null` reflect explicitly and only this state; in every case where we've observed the memo (because we created the transaction, or decrypted it) it should be identical to the memo data that's encoded into the transaction itself.

Reply:
(Remember that the `sent_notes` table is badly named at this point; it also contains metadata for transparent outputs.)

Reply:
If that is the desired change, then this should be done with a database migration. The current `null` value is aliasing "implicit empty memo" (which previously was the only state this column stored) with "unknown memo" (present only since your change to store received change notes in the `sent_notes` table). To properly fix this, we need a database migration that checks whether the `raw` field is non-null (indicating we do have the transaction data populated), and decrypts the outputs to re-fill the memo column.

Reply:
Getting rid of these assumptions would be much appreciated by our future selves. 

Maybe it is too much of a lift, or out of scope for this PR?

It feels much safer to actually keep the things an they are with the current conventions ( extra docs will help for sure ) and keep the PR more concise.

Reply:
`memo_repr` is undocumented, so I went digging to determine its in-practice definition. Prior to this PR, the only two places we call `memo_repr` are:
- `put_received_note` using `ReceivedSaplingOutput::memo`, which is undocumented but in practice for our impls `None` means "we don't have the memo data".
- `put_sent_output` using a provided `Option<&MemoBytes>`. This is only called inside `store_decrypted_tx`: once with `Some(&memo)` for decrypted Sapling outputs, and once with `None` for transparent outputs.

So just from this we already have two definitions of `None` as the output of `memo_repr`, but they happen to not be aliased as they are used in different tables. The `memo` column in the `sapling_received_notes` table uses `null` in this context to mean "unknown memo", while the `memo` column in the `sent_notes` table uses `null` in this context to mean "not applicable to this output pool".

This PR alters `insert_sent_output` to use `memo_repr` as well, replacing code that is equivalent to the previous implementation of `memo_repr`. The field there is `SentTransactionOutput::memo`, which per the other comment thread uses `None` to mean either "transparent output" or "no change memo specified" as of this PR, but previously also used it to mean "unknown memo". So AFAICT if the "no change memo specified" case is fixed per my new comment below, then the `sent_notes` table's `memo` column has a single definition of `null` after this PR.